### PR TITLE
Path of truth fixes

### PR DESCRIPTION
**Pull request by fahslaj**

### CODE REVIEW

File: src/main/java/cceclipseplugin/core/PluginManager.java [hunk 1]
```diff
@@ -387,12 +385,12 @@ private void registerNotificationHooks() {
 				return;
 			}
 			Project project = dataManager.getSessionStorage().getProjectById(mm.getProjectIDForFileID(resId));
-			String pathToFile = meta.getFilePath();
 			IProject p = ResourcesPlugin.getWorkspace().getRoot().getProject(project.getName());
-			IFile file = p.getFile(pathToFile);
+			IFile file = p.getFile(meta.getFilePath());
+			String fullPath = file.getFullPath().toString();
```
Comment (fahslaj) [inline src/main/java/cceclipseplugin/core/PluginManager.java]: Refactor this variable to specify what type of path this is (workspace relative, project relative with or without file name, or absolute)

File: src/main/java/cceclipseplugin/core/CCIgnore.java [hunk 2]
```diff
@@ -38,14 +41,15 @@
...
-		IFile file = p.getFile(new Path(Paths.get(FILENAME).normalize().toString()));
+		IFile file = p.getFile(new Path(FILENAME));
 		
 		if (!file.exists()) {
 			InputStream in = new ByteArrayInputStream(DEFAULT_CONTENTS.getBytes());
 			
 			try {
 				// warn directory watching
-				PluginManager.getInstance().putFileInWarnList(file.getProjectRelativePath().toString(), FileCreateResponse.class);
+				IPath workspaceRelativePath = p.getFullPath().append(file.getProjectRelativePath());
```
Comment (loganga) [inline src/main/java/cceclipseplugin/core/CCIgnore.java]: Use file.getFullPath() instead, should return the workspace relative path for the file

File: src/main/java/cceclipseplugin/core/PluginManager.java [hunk 3]
```diff
@@ -405,32 +403,33 @@ private void registerNotificationHooks() {
...
-				System.out.println("Tried to delete file that does not exist: " + pathToFile);
+				System.out.println("Tried to delete file that does not exist: " + fullPath);
 			}
 		});
 		// File.Change
 		wsManager.registerNotificationHandler("File", "Change", dataManager.getPatchManager());
 	}
 		
-	private boolean renameFile(IFile file, IPath newPath, long projId) {
+	private boolean renameFile(IFile file, IPath newFullPath, long projId) {
```
Comment (fahslaj) [inline src/main/java/cceclipseplugin/core/PluginManager.java]: Same here; specify what type of path this is with the variable name. (Full path is ambiguous)

File: src/main/java/cceclipseplugin/core/EclipseRequestManager.java [hunk 4]
```diff
@@ -116,7 +114,8 @@ public void pullFileAndCreate(IProject p, Project ccp, File file, IProgressMonit
 							IFolder newFolder = p.getFolder(currentFolder);
 							try {
 								if (!newFolder.exists()) {
-									pm.putFileInWarnList(relPath.toString(), FileCreateNotification.class);
+									IPath workspaceRelativePath = p.getFullPath().append(relPath);
```
Comment (loganga) [inline src/main/java/cceclipseplugin/core/EclipseRequestManager.java]: Make sure workspaceRelativePath has the filename in it; line 104 is probably where this should be done

File: src/main/java/cceclipseplugin/editor/DocumentManager.java [hunk 5]
```diff
@@ -184,16 +184,20 @@ public void handleNotification(Notification n) {
...
-
-		this.applyPatch(n.getResourceID(), filepath, patches);
-
+		
 		ProjectMetadata projMeta = PluginManager.getInstance().getMetadataManager()
-				.getProjectMetadata(ResourcesPlugin.getWorkspace().getRoot().getLocation().toString() + "/");
+				.getProjectMetadata(projectRootPath.toString());
+		
+		IPath fileRelativePathWithName = new Path(fileMeta.getFilePath());
+		IPath projectRelativePath = new Path('/' + projMeta.getName());
```
Comment (fahslaj) [inline src/main/java/cceclipseplugin/editor/DocumentManager.java]: Does this need to have a preceding / ? or will the IPath constructor add it automatically.

File: src/main/java/cceclipseplugin/editor/DocumentManager.java [hunk 6]
```diff
@@ -73,7 +73,7 @@ public void setCurrFile(String filePath) {
 			this.currFile = null;
 			return;
 		}
-		this.currFile = Paths.get(filePath).toString();
+		this.currFile = filePath;
```
Comment (loganga) [inline src/main/java/cceclipseplugin/editor/DocumentManager.java]: Probably should update Java docs to more explicitly denote what kind of paths should be used in this instance (absolute paths)

File: src/main/java/cceclipseplugin/editor/listeners/DirectoryListener.java [hunk 7]
```diff
@@ -86,15 +86,15 @@ protected void handleFile(IResourceDelta delta) {
 			if ((delta.getFlags() & IResourceDelta.MOVED_TO) != 0) {
 				
 				IPath relativeMovedToPath = delta.getMovedToPath().removeFirstSegments(1);
-				
+				// replace \\ with /?
```
Comment (fahslaj) [inline src/main/java/cceclipseplugin/editor/listeners/DirectoryListener.java]: This isn't necessary.

File: src/main/java/cceclipseplugin/editor/DocumentManager.java [hunk 8]
```diff
@@ -204,18 +208,18 @@ public void handleNotification(Notification n) {
...
 	 * @param fileId
 	 *            fileId to patch; this is mainly used for passing to clientCore
-	 * @param filePath
+	 * @param absolutePath
 	 *            absolute file path; used as key in editorMap, and patches.
 	 * @param patches
 	 *            the list of patches to apply, in order.
 	 */
-	public void applyPatch(long fileId, String filePath, List<Patch> patches) {
+	public void applyPatch(long fileId, String absolutePath, String workspaceRelativePath, List<Patch> patches) {
```
Comment (loganga) [inline src/main/java/cceclipseplugin/editor/DocumentManager.java:218]: Make sure include a javadoc for the workspaceRelativePath parameter that specifies that you need a path that includes the filename

File: src/main/java/cceclipseplugin/editor/listeners/DirectoryListener.java [hunk 9]
```diff
@@ -77,7 +77,7 @@ protected void handleFile(IResourceDelta delta) {
 		PluginManager pm = PluginManager.getInstance();
 		MetadataManager mm = pm.getMetadataManager();
 		FileMetadata fileMeta = mm.getFileMetadata(f.getFullPath().toString());
-		String path = f.getProjectRelativePath().toString();
+		String workspaceRelativePath = f.getProjectRelativePath().toString();
```
Comment (loganga) [inline src/main/java/cceclipseplugin/editor/listeners/DirectoryListener.java]: getProjectRelativePath() doesn't return the workspace-relative path, getFullPath() does

File: src/main/java/cceclipseplugin/editor/listeners/DirectoryListener.java (hunk 7 above)
Comment (loganga) [inline src/main/java/cceclipseplugin/editor/listeners/DirectoryListener.java]: ¯\\\_(ツ)\_/¯

File: src/main/java/cceclipseplugin/editor/DocumentManager.java (hunk 8 above)
Comment (fahslaj) [reply src/main/java/cceclipseplugin/editor/DocumentManager.java:218]: Fixed.

**Comment by loganga** [comment]
Looks fly 2 me 👌

**Pull request merged**